server/py/services/api/initial_data.py

```diff
@@ -877,40 +877,84 @@ def _perform_version_8_data_migrations(
     db: framework.db.sqldb.db.SQLDB, db_session: sqlalchemy.orm.Session
 ):
-    _migrate_function_kind(db, db_session, chunk_size=500)
+    _migrate_data(
+        db,
+        db_session,
+        framework.db.sqldb.models.Function,
+        "kind",
+        _handle_function_kind,
+    )
+    _migrate_data(
+        db,
+        db_session,
+        framework.db.sqldb.models.ArtifactV2,
+        "producer_uri",
+        _handle_artifact_producer_uri,
+    )
+
 
+def _handle_function_kind(records: list, record):
+    function_dict = record.struct
+    record.kind = function_dict.pop("kind", None)
+    if record.kind is not None:
+        record.struct = function_dict
+        records.append(record)
 
-def _migrate_function_kind(
-    db: framework.db.sqldb.db.SQLDB, db_session: sqlalchemy.orm.Session, chunk_size: int
+
+def _handle_artifact_producer_uri(records: list, record):
+    record.producer_uri = (
+        record.full_object.get("spec", {}).get("producer", {}).get("uri", None)
+    )
+    if record.producer_uri is not None:
+        records.append(record)
+
+
+def _migrate_data(
+    db: framework.db.sqldb.db.SQLDB,
+    db_session: sqlalchemy.orm.Session,
+    model,
+    field_name: str,
+    handle_field_record_func,
+    chunk_size: int = 500,
 ):
     offset = 0
-    number_of_functions_to_migrate = (
-        db._query(db_session, framework.db.sqldb.models.Function)
-        .filter(framework.db.sqldb.models.Function.kind.is_(None))
+    # Get count of records to migrate
+    count = (
+        db._query(db_session, model)
+        .filter(getattr(model, field_name).is_(None))
         .count()
     )
-    logger.info(f"Number of functions to migrate: {number_of_functions_to_migrate}")
+    logger.info(f"Number of {model.__name__.lower()}s to migrate: {count}")
+
     while True:
-        logger.info("Fetching functions to migrate", offset=offset)
-        q = db._query(db_session, framework.db.sqldb.models.Function)
-        q = q.order_by(framework.db.sqldb.models.Function.id)
-        q = q.limit(chunk_size)
-        q = q.offset(offset)
-        functions = []
-        for function in q.all():
-            function_dict = function.struct
-            function.kind = function_dict.pop("kind", None)
-            if function.kind is None:
-                continue
-            function.struct = function_dict
-            functions.append(function)
-        if functions:
-            logger.info("Committing migrated functions", count=len(functions))
-            db_session.add_all(functions)
-            db._commit(db_session, functions)
-        if len(functions) < chunk_size:
-            logger.info("No more functions to migrate")
+        logger.info(f"Fetching {model.__name__.lower()}s to migrate", offset=offset)
+
+        # Build query to fetch the data in chunks
+        query = (
+            db._query(db_session, model)
+            .order_by(getattr(model, "id"))
+            .limit(chunk_size)
```

Comment:
Don't we want to filter by `model.field_name != None`? Since this is chunked, it's possible that migration would go half-way and then the service is restarted. In this case you want to only work on records that weren't already migrated.
Even without restart, doing it this way means you won't need `.offset()` anymore since any new query starts from offset 0 (anything that was already updated is not returned by the BE), and this actually make the query more efficient overall.